### PR TITLE
react-scripts@feature/core

### LINKED ISSUE #17: Core Redux component from function to class
As Fela component was rewritten into class component to avoid re-rendering styles, wouldn't it be better to do the same for Redux? There shouldnt be a problem since it is one of the top level components and never re-renders but just to be sure.

I mean [this code](https://github.com/AckeeCZ/create-react-app/blob/master/packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx).

### LINKED ISSUE #29: remove config for core - divide it and move it to single core sub-modules

### LINKED ISSUE #34: Fix import of loadable from @ackee/lucas
In [`/packages/react-scripts/template/src/HOC/loadable.jsx`](https://github.com/AckeeCZ/create-react-app/blob/master/packages/react-scripts/template/src/HOC/loadable.jsx) there is an import of `loadable` factory from @ackee/lucas. The problem is, it is defined as a *default* import, but @ackee/lucas doesn't have any default exports, fix it please 🙂

### PR DESCRIPTION
**Pull request by cermakjiri**
Doporucuju se na upraveny `core` modulu podivat primo lokalne. Podle me to bude prehlednejsi, nez koukat na ty diffy.

## Major changes
- rozdelil jsem globalni konfiguraci predtim umistenou `index.js` do odpovidajicich `core` sub-modulu
- globalní `localization` jsem přesunul přímo do `core/modules/localization`
- nejvetsi zmeny jsem provedl u `core/modules/redux` a `core/modules/fela`, tak na ne koukne obzvlast pozorne
- `core/modules/fela` nyni obsahuje i `styles` slozku. Dava mi smysl mit nadefinovane fela `theme` primo zde, nez ho tahat z globalnich `styles`. Stejne tak staticke styly, ktere se do fely napoji jednou pri `componentDidMount`. 

---

## Minor changes
- pouzil jsem `requestIdleCallback` pro initializaci sentry
- odstranil jsem pozustatky `components/FormFields` a pouzil radsi komponenty z `@ackee/mateus`
- oprava importu `loadable` (resolves #34)
- aktualizace READMEs, aby sedely nove core module strukture
- odstraneni `routingHistory` reduceru a sagy (pouziva se ve vyjimecnych pripadech, developer si ji prida sam pokud bude potreba)
---

Samozrejme komentujte dle libosti, nicmene jestli je to na delsi diskuzi, probral bych to radeji na nasledujicim meetingu. 

---

resolves #17, resolves #29.

### CODE REVIEW

File: packages/react-scripts/template/src/modules/core/modules/sentry/config/index.js [hunk 1]
```diff
@@ -24,7 +24,9 @@ if (!options.dsn) {
 }
 
 export const initializeSentry = () => {
-    Sentry.init(options);
+    const init = () => Sentry.init(options);
+
+    window.requestIdleCallback ? window.requestIdleCallback(init) : init();
```
Comment (horaklukas) [inline packages/react-scripts/template/src/modules/core/modules/sentry/config/index.js]: Didn't know 👍

File: packages/react-scripts/template/src/services/sagas/index.js [hunk 2]
```diff
@@ -1,23 +1,13 @@
...
-/**
- * Root generator for all application sagas
- */
-export default function*() {
+export default function* appSaga() {
     yield all([
-        httpClient(),
         modules(),
-        formSaga(),
-        routingHistory.saga(),
```
Comment (horaklukas) [inline packages/react-scripts/template/src/services/sagas/index.js]: Same as with reducers, how do you determine, which sagas are moved to `core/modules/redux/services/saga` and which stay here? Explain it please 🙂

File: packages/react-scripts/template/src/index.jsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import Core from 'Modules/core';
+import Application from 'Modules/application';
+
+ReactDOM.render(
+    <Core>
+        <Application />
+    </Core>,
+    document.getElementById('root'),
+);
```
Comment (horaklukas) [inline packages/react-scripts/template/src/index.jsx]: Where did the application loader go? 🙂

File: packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx [hunk 4]
```diff
@@ -1,29 +1,23 @@
...
-Redux.propTypes = {
-    children: PropTypes.node.isRequired,
-    // eslint-disable-next-line
-    initialState: PropTypes.object,
-    // eslint-disable-next-line
-    reducers: PropTypes.object.isRequired,
-    saga: PropTypes.func,
-    middlewares: PropTypes.arrayOf(PropTypes.func),
-};
+        this.store = configureStore();
```
Comment (horaklukas) [inline packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx]: That's much better 👍

File: packages/react-scripts/template/src/modules/core/modules/redux/dependencies.js [hunk 5]
```diff
@@ -7,10 +7,16 @@ export { createStore, applyMiddleware, combineReducers, compose } from 'redux';
...
 
 export * as Consts from 'Consts';
 export { default as config } from 'Config';
 
 export { routerMiddlewareWithHistory, connectRouterWithHistory } from '../router';
 export { reducer as translateReducer, saga as localizationSaga } from '../localization';
 export { saga as sentrySaga, sentryMiddleware } from '../sentry';
+
+export { default as reducers } from 'Services/reducers';
+export { default as saga } from 'Services/sagas';
```
Comment (horaklukas) [inline packages/react-scripts/template/src/modules/core/modules/redux/dependencies.js]: If I understand it right, here will be import of all reducers/sagas from other modules, right? I can't get rid of feeling it will lead to circular dependencies (but maybe it's unnecessary fear) between this and other modules..

File: packages/react-scripts/template/src/index.jsx (hunk 3 above)
Comment (cermakjiri) [reply packages/react-scripts/template/src/index.jsx]: I've removed code-splitting for the `application` module for the following reason:

With the code-splitting:
```
  660.61 KB  build/static/js/2.92cf8ac6.chunk.js
  200.84 KB  build/static/js/main.3525ba4f.chunk.js
  2.89 KB    build/static/js/3.8eb6bdc1.chunk.js (the application module)
  1.15 KB    build/static/js/runtime~main.da4b5e3c.js
```

After the code-splitting:
```
  660.43 KB (-179 B)  build/static/js/2.894dce0c.chunk.js
  200.65 KB (-198 B)  build/static/js/main.e98227c0.chunk.js
  762 B (-412 B)      build/static/js/runtime~main.b3dac7a9.js
```

That's `expenses` production build of the current version.

Until we will find a way to properly do the code-splitting, it seems obvious to leave it in this case completely.

File: packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx (hunk 4 above)
Comment (cermakjiri) [reply packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx]: Right? I am very happy with the code complexity reduction that has been achieved. 🙂

File: packages/react-scripts/template/src/services/reducers/index.js [hunk 6]
```diff
@@ -10,7 +9,6 @@ const reducers = {
     // @use-auth-module-begin
     auth,
```
[the remark this answers is not in the document]
Comment (cermakjiri) [reply packages/react-scripts/template/src/services/reducers/index.js]: I am glad you've mentioned this issue. I wasn't sure at the time of creating the MR what's the best solution. 

The benefit of importing reducers/sagas in global services is that it might be imported directly from a module and it hasn't to be imported through `dependencies`. However, it would make more sense to import them directly in `core/redux`.

I can't see the issue with circular dep.:
`module/foo` -> `core/modules/redux` -> `core/components/Core` -> `index.jsx` 
In other words, everything goes into the `redux` module (not from it) and the `redux` module is only used in `Core` component. So we should be fine.

[the remark this answers is not in the document]
Comment (cermakjiri) [reply packages/react-scripts/template/src/services/reducers/index.js]: Since there wouldn't been a need for global services, shouldn't be done the same for global `components`, `containers` and `HOC`?

File: packages/react-scripts/template/src/modules/core/modules/fela/components/Fela.jsx [hunk 7]
```diff
@@ -1,46 +1,31 @@
...
     constructor(props) {
         super(props);
 
-        this.renderer = createRenderer(config);
+        this.renderer = createRenderer(Config.rendererConfig);
     }
 
     componentDidMount() {
-        const { staticCSS, reset, fonts } = this.props;
+        applyStaticCSS(this.renderer, Config.staticCSS);
```
Comment (jancama2) [inline packages/react-scripts/template/src/modules/core/modules/fela/components/Fela.jsx]: I thought these will come from global styles, I do understand why you have put it in here, but I think the parts of the app that are going to change more often should be in global scope (global styles, reducer and sagas from modules etc.) so you dont have to go into core every time you want to wire in something. Core, for me, is something stale and changing rarely.

Comment (jancama2) [reply packages/react-scripts/template/src/modules/core/modules/fela/components/Fela.jsx]: This applies for fonts too.

File: packages/react-scripts/template/src/modules/core/modules/fela/config/index.js [hunk 8]
```diff
@@ -1,49 +1,22 @@
...
-    devMode: isNotProd,
+export const rendererConfig = {
+    devMode: Consts.isEnvDevelopment,
     enhancers,
     plugins,
+    theme,
 };
 
-export default config;
+export const staticCSS = [resets, vendors];
```
Comment (jancama2) [inline packages/react-scripts/template/src/modules/core/modules/fela/config/index.js]: As I've mentioned above, would like to see these in global scope.

File: packages/react-scripts/template/src/modules/core/modules/fela/styles/index.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export * as theme from './theme';
+export { default as vendors } from './vendors.less';
```
Comment (jancama2) [inline packages/react-scripts/template/src/modules/core/modules/fela/styles/index.js]: I dont think styles should be here. If, for some reason, I dont want fela in the project, I have to restructure the project. If fela take it from the global styles folder, you can't get rid of fela easier.

Comment (jancama2) [reply packages/react-scripts/template/src/modules/core/modules/fela/styles/index.js]: This doesnt apply for theme which is really just fela dep.

File: packages/react-scripts/template/src/modules/core/modules/localization/translations/cs.yml [hunk 10]
```diff
@@ -24,4 +24,4 @@ page.auth.title: 'Auth'
 page.noMatch.title: 'No match'
```
Comment (jancama2) [inline packages/react-scripts/template/src/modules/core/modules/localization/translations/cs.yml]: Again, not sure if this should be here, it is going to change rapidly and I dont think you should go to core module that often. With this logic, almost everything can fit in the core module.

**I think we should brainstorm core responsibilities before approving the PR.** @horaklukas thoughts?

File: packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx [hunk 11]
```diff
@@ -1,29 +1,23 @@
...
-    initialState: {},
-    *saga() {
-        yield;
-    },
-    middlewares: [],
-};
+    render() {
+        const { store } = this;
+
+        return store ? <Provider store={store}>{this.props.children}</Provider> : null;
```
Comment (jancama2) [inline packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx]: do you really need to check it?

Comment (jancama2) [reply packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx]: and I think you should at least throw an error if there is no store, but I think it can't happen to have no store here.

File: packages/react-scripts/template/src/modules/core/modules/redux/config/enhancer.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            Sentry.captureException(e);
+        },
+    });
+
+    const middlewares = [routerMiddlewareWithHistory, sagaMiddleware, sentryMiddleware];
+    const enhancerArgs = [applyMiddleware(...middlewares)];
+
+    if (config.devTools) {
+        // eslint-disable-next-line
+        const reduxDevTools = window.__REDUX_DEVTOOLS_EXTENSION__;
```
Comment (jancama2) [inline packages/react-scripts/template/src/modules/core/modules/redux/config/enhancer.js]: I have noticed that redux tools are working even for prod build. This should not happen.

File: packages/react-scripts/template/src/modules/core/modules/redux/dependencies.js (hunk 5 above)
Comment (jancama2) [reply packages/react-scripts/template/src/modules/core/modules/redux/dependencies.js]: From my point of view, core should be stand alone and shouldnt be used by other modules. If so, then it is badly implemented. Having reducers and sagas from other modules wired in global `Services` and then injected in here is the only good way of doing it, for me.

File: packages/react-scripts/template/src/services/reducers/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jancama2) [reply packages/react-scripts/template/src/services/reducers/index.js]: We surely need `src/services/reducers/index.js` (and `src/services/sagas/index.js`) as it is a wiring point for reducers from other modules but the core. This wiring point then should be injected into the core. You dont want to go to the core every time you want to add new reducer,saga etc.

It may look empty not, but take a look at this file from my other project:
```
import { reducer as auth } from 'Modules/auth';
import { reducer as secrets } from 'Modules/secrets';

import entities from './entities';
import api from './api';

// NOTE:
// Do not use combineReducers() here,
// export reducers as plain object, not as a function.
const reducers = {
    auth,
    secrets,
    entities,
    api,
};

export default reducers;
```

File: packages/react-scripts/template/src/services/sagas/index.js (hunk 2 above)
Comment (jancama2) [reply packages/react-scripts/template/src/services/sagas/index.js]: Sagas from external dependencies are in the core and app sagas are wired in here (mainly sagas from the other modules but the core module. Again, some example from my other project:
```
import { all } from 'redux-saga/effects';

import { saga as auth } from 'Modules/auth';
import { saga as secrets } from 'Modules/secrets';
import { saga as groups } from 'Modules/groups';
import { saga as whitelist } from 'Modules/whitelist';

export default function*() {
    yield all([auth(), secrets(), groups(), whitelist()]);
}
```

I don't like the `modules` saga here thought. They should be imported in here, as in my example.

Comment (jancama2) [reply packages/react-scripts/template/src/services/sagas/index.js]: Generaly, the core module is a wiring point of all the external deps and it shouldn't do nothing else then that. I don't want to go into the core every time I add some new module or other functionality to the app. The core should be stale and change just if you need to change the architecture of the application (for example you want to use MobX instead of Redux). 

As I've mentioned above. We should braingstorm responsibilities of the core before approving this PR.

File: packages/react-scripts/template/src/services/reducers/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jancama2) [reply packages/react-scripts/template/src/services/reducers/index.js]: > Since there wouldn't be a need for global `services`, shouldn't be done the same for global `components`, `containers` and `HOC`?

There is kinda no need, but I wouldn't like to see it in the core module. Maybe "layout" module or something else.

File: packages/react-scripts/template/src/index.jsx (hunk 3 above)
Comment (horaklukas) [reply packages/react-scripts/template/src/index.jsx]: so since there is no code splitting, we don't need Application laoder?

**Comment by horaklukas** [comment]
Generally, I like and agree with @jancama2 approach to core module 👍 The things that often change would be at "global" level while core module contained default configuration that changes rarely.

File: packages/react-scripts/template/src/services/reducers/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (cermakjiri) [reply packages/react-scripts/template/src/services/reducers/index.js]: > > Since there wouldn't be a need for global `services`, shouldn't be done the same for global `components`, `containers` and `HOC`?
> 
> There is kinda no need, but I wouldn't like to see it in the core module. Maybe "layout" module or something else.

I should express better myself. Definitely, those entities shouldn't be placed at the `core` module. That's nonsense. But they could placed, as your're saying, in `layout` module or maybe  it'd be called `ui` module.

File: packages/react-scripts/template/src/index.jsx (hunk 3 above)
Comment (cermakjiri) [reply packages/react-scripts/template/src/index.jsx]: > so since there is no code splitting, we don't need Application laoder?

That's true, I'll remove it.

File: packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx (hunk 11 above)
Comment (cermakjiri) [reply packages/react-scripts/template/src/modules/core/modules/redux/components/Redux.jsx]: You're right, there's no point in checking it. I was creating the store initially in `componentDidMount` and forgot to remove the condition.

File: packages/react-scripts/template/src/modules/core/modules/redux/config/enhancer.js (hunk 12 above)
Comment (cermakjiri) [reply packages/react-scripts/template/src/modules/core/modules/redux/config/enhancer.js]: Are you sure? In the config is following condition: `process.env.NODE_ENV !== Consts.env.PRODUCTION`

Comment (cermakjiri) [reply packages/react-scripts/template/src/modules/core/modules/redux/config/enhancer.js]: I couldn't sucessfully replicated. I can only see redux devtools instances of other projects.

File: packages/react-scripts/template/src/modules/core/modules/fela/styles/index.js (hunk 9 above)
Comment (cermakjiri) [reply packages/react-scripts/template/src/modules/core/modules/fela/styles/index.js]: That's a fair point, styles should be moved back to the global level.
